azurerm/internal/services/web/app_service.go

```diff
@@ -1909,7 +1913,28 @@ func expandAppServiceIpRestriction(input interface{}) ([]web.IPSecurityRestricti
 		priority := restriction["priority"].(int)
 		action := restriction["action"].(string)
 
-		if vNetSubnetID != "" && ipAddress != "" && serviceTag != "" {
+		if d.HasChange(iterKey + ".ip_address") {
+			if ipAddress != "" {
+				vNetSubnetID = ""
+				serviceTag = ""
+			}
+		}
+
+		if d.HasChange(iterKey + ".service_tag") {
+			if serviceTag != "" {
+				vNetSubnetID = ""
+				ipAddress = ""
+			}
+		}
+
+		if d.HasChanges(iterKey+".subnet_id", iterKey+".virtual_network_subnet_id") {
+			if vNetSubnetID != "" {
+				ipAddress = ""
+				serviceTag = ""
+			}
+		}
+
+		if (vNetSubnetID != "" && (ipAddress != "" || serviceTag != "")) || (ipAddress != "" && (vNetSubnetID != "" || serviceTag != "")) {
```

Comment:
Personally I'd extract this if-statement to a variable named after what it's checking. 

e.g

```
validAddress := vNetSubnetID != "" && ..

if validAddress {
    // code here 
}
```